{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68a13ecd",
   "metadata": {},
   "source": [
    "# Tutorial: Using Cohere with Elasticsearch\n",
    "\n",
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/elastic/elasticsearch-labs/blob/main/notebooks/integrations/cohere/cohere-elasticsearch.ipynb)\n",
    "\n",
    "This tutorial shows you how to compute embeddings with\n",
    "Cohere using the inference API and store them for efficient vector or hybrid\n",
    "search in Elasticsearch. This tutorial uses the Python Elasticsearch client\n",
    "to perform the operations.\n",
    "\n",
    "You'll learn how to:\n",
    "* create an inference endpoint for text embedding using the Cohere service,\n",
    "* create the necessary index mapping for the Elasticsearch index,\n",
    "* build an inference pipeline to ingest documents into the index together with the embeddings,\n",
    "* perform hybrid search on the data,\n",
    "* rerank search results by using Cohere's rerank model,\n",
    "* design a RAG system with Cohere's Chat API.\n",
    "\n",
    "The tutorial uses the [SciFact](https://huggingface.co/datasets/mteb/scifact) data set.\n",
    "\n",
    "Refer to [Cohere's tutorial](https://docs.cohere.com/docs/elasticsearch-and-cohere) for an example using a different data set."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f068b1b",
   "metadata": {},
   "source": [
    "## ðŸ§° Requirements\n",
    "\n",
    "For this example, you will need:\n",
    "\n",
    "- An Elastic deployment with minimum **4GB machine learning node**\n",
    "   - We'll be using [Elastic Cloud](https://www.elastic.co/guide/en/cloud/current/ec-getting-started.html) for this example (available with a [free trial](https://cloud.elastic.co/registration?utm_source=github&utm_content=elasticsearch-labs-notebook))\n",
    "   \n",
    "- A paid [Cohere account](https://cohere.com/) is required to use the Inference API with \n",
    "the Cohere service as the Cohere free trial API usage is limited.\n",
    "\n",
    "- Python 3.7 or later."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aac02d37",
   "metadata": {},
   "source": [
    "## Install and import required packages\n",
    "\n",
    "Install Elasticsearch and Cohere:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "50940b4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking in indexes: http://mirrors.aliyun.com/pypi/simple/\n",
      "Requirement already satisfied: elasticsearch in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (8.14.0)\n",
      "Requirement already satisfied: elastic-transport<9,>=8.13 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from elasticsearch) (8.13.0)\n",
      "Requirement already satisfied: urllib3<3,>=1.26.2 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from elastic-transport<9,>=8.13->elasticsearch) (2.2.2)\n",
      "Requirement already satisfied: certifi in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from elastic-transport<9,>=8.13->elasticsearch) (2024.2.2)\n",
      "Looking in indexes: http://mirrors.aliyun.com/pypi/simple/\n",
      "Requirement already satisfied: cohere in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (5.5.8)\n",
      "Requirement already satisfied: boto3<2.0.0,>=1.34.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (1.34.131)\n",
      "Requirement already satisfied: fastavro<2.0.0,>=1.9.4 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (1.9.4)\n",
      "Requirement already satisfied: httpx>=0.21.2 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (0.27.0)\n",
      "Requirement already satisfied: httpx-sse<0.5.0,>=0.4.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (0.4.0)\n",
      "Requirement already satisfied: parameterized<0.10.0,>=0.9.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (0.9.0)\n",
      "Requirement already satisfied: pydantic>=1.9.2 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (2.6.3)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (2.31.0)\n",
      "Requirement already satisfied: tokenizers<1,>=0.15 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (0.15.2)\n",
      "Requirement already satisfied: types-requests<3.0.0,>=2.0.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (2.32.0.20240622)\n",
      "Requirement already satisfied: typing_extensions>=4.0.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from cohere) (4.9.0)\n",
      "Requirement already satisfied: botocore<1.35.0,>=1.34.131 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from boto3<2.0.0,>=1.34.0->cohere) (1.34.131)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from boto3<2.0.0,>=1.34.0->cohere) (1.0.1)\n",
      "Requirement already satisfied: s3transfer<0.11.0,>=0.10.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from boto3<2.0.0,>=1.34.0->cohere) (0.10.1)\n",
      "Requirement already satisfied: anyio in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from httpx>=0.21.2->cohere) (4.3.0)\n",
      "Requirement already satisfied: certifi in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from httpx>=0.21.2->cohere) (2024.2.2)\n",
      "Requirement already satisfied: httpcore==1.* in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from httpx>=0.21.2->cohere) (1.0.4)\n",
      "Requirement already satisfied: idna in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from httpx>=0.21.2->cohere) (3.6)\n",
      "Requirement already satisfied: sniffio in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from httpx>=0.21.2->cohere) (1.3.1)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from httpcore==1.*->httpx>=0.21.2->cohere) (0.14.0)\n",
      "Requirement already satisfied: annotated-types>=0.4.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from pydantic>=1.9.2->cohere) (0.6.0)\n",
      "Requirement already satisfied: pydantic-core==2.16.3 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from pydantic>=1.9.2->cohere) (2.16.3)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests<3.0.0,>=2.0.0->cohere) (3.3.2)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests<3.0.0,>=2.0.0->cohere) (2.2.2)\n",
      "Requirement already satisfied: huggingface_hub<1.0,>=0.16.4 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from tokenizers<1,>=0.15->cohere) (0.20.3)\n",
      "Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from botocore<1.35.0,>=1.34.131->boto3<2.0.0,>=1.34.0->cohere) (2.8.2)\n",
      "Requirement already satisfied: filelock in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from huggingface_hub<1.0,>=0.16.4->tokenizers<1,>=0.15->cohere) (3.13.1)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from huggingface_hub<1.0,>=0.16.4->tokenizers<1,>=0.15->cohere) (2024.2.0)\n",
      "Requirement already satisfied: tqdm>=4.42.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from huggingface_hub<1.0,>=0.16.4->tokenizers<1,>=0.15->cohere) (4.66.2)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from huggingface_hub<1.0,>=0.16.4->tokenizers<1,>=0.15->cohere) (6.0.1)\n",
      "Requirement already satisfied: packaging>=20.9 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from huggingface_hub<1.0,>=0.16.4->tokenizers<1,>=0.15->cohere) (23.2)\n",
      "Requirement already satisfied: six>=1.5 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.35.0,>=1.34.131->boto3<2.0.0,>=1.34.0->cohere) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install elasticsearch\n",
    "!pip install cohere"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b24d95b",
   "metadata": {},
   "source": [
    "Import the required packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1a5a5eeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from elasticsearch import Elasticsearch, helpers\n",
    "import cohere\n",
    "import json\n",
    "import requests\n",
    "from dotenv import load_dotenv\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d0dacc5",
   "metadata": {},
   "source": [
    "## Create an Elasticsearch client"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74841401",
   "metadata": {},
   "source": [
    "Now you can instantiate the Python Elasticsearch client.\n",
    "\n",
    "Then create a `client` object that instantiates an instance of the `Elasticsearch` class."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "76394974",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://elastic:uK+7WbkeXMzwk9YvP-H3@localhost:9200\n",
      "{'name': 'liuxgm.local', 'cluster_name': 'elasticsearch', 'cluster_uuid': 'coIKHIPsTf2_aWWQ8TO4bw', 'version': {'number': '8.14.1', 'build_flavor': 'default', 'build_type': 'tar', 'build_hash': '93a57a1a76f556d8aee6a90d1a95b06187501310', 'build_date': '2024-06-10T23:35:17.114581191Z', 'build_snapshot': False, 'lucene_version': '9.10.0', 'minimum_wire_compatibility_version': '7.17.0', 'minimum_index_compatibility_version': '7.0.0'}, 'tagline': 'You Know, for Search'}\n"
     ]
    }
   ],
   "source": [
    "load_dotenv()\n",
    " \n",
    "ES_USER = os.getenv(\"ES_USER\")\n",
    "ES_PASSWORD = os.getenv(\"ES_PASSWORD\")\n",
    "ES_ENDPOINT = os.getenv(\"ES_ENDPOINT\")\n",
    "COHERE_API_KEY = os.getenv(\"COHERE_API_KEY\")\n",
    " \n",
    "url = f\"https://{ES_USER}:{ES_PASSWORD}@{ES_ENDPOINT}:9200\"\n",
    "print(url)\n",
    " \n",
    "client = Elasticsearch(url, ca_certs = \"./http_ca.crt\", verify_certs = True)\n",
    "print(client.info())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d29b809",
   "metadata": {},
   "source": [
    "## Create the inference endpoint\n",
    "\n",
    "Create the inference endpoint first. In this example, the inference endpoint \n",
    "uses Cohere's `embed-english-v3.0` model and the `embedding_type` is set to\n",
    "`byte`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ae8ae88c",
   "metadata": {},
   "outputs": [],
   "source": [
    "co = cohere.Client(api_key=COHERE_API_KEY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6e524ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from elasticsearch import BadRequestError\n",
    "\n",
    "client.inference.delete_model(inference_id=\"cohere_embeddings\")\n",
    "\n",
    "try: \n",
    "    client.inference.put_model(\n",
    "        task_type=\"text_embedding\",\n",
    "        inference_id=\"cohere_embeddings\",\n",
    "        body={\n",
    "            \"service\": \"cohere\",\n",
    "            \"service_settings\": {\n",
    "                \"api_key\": COHERE_API_KEY,\n",
    "                \"model_id\": \"embed-english-v3.0\",\n",
    "                \"embedding_type\": \"byte\",\n",
    "            },\n",
    "        },\n",
    "    )\n",
    "except BadRequestError as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "884e424d",
   "metadata": {},
   "source": [
    "You can find your API keys in your Cohere dashboard under the\n",
    "[API keys section](https://dashboard.cohere.com/api-keys)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "178c32db",
   "metadata": {},
   "source": [
    "## Create the index mapping\n",
    "\n",
    "Create the index mapping for the index that will contain the embeddings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "35ab26b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "index_name=\"cohere-embeddings\"\n",
    "\n",
    "client.indices.delete(index=index_name)\n",
    "\n",
    "if not client.indices.exists(index=index_name):\n",
    "    client.indices.create(\n",
    "        index=index_name,\n",
    "        settings={\"index\": {\"default_pipeline\": \"cohere_embeddings\"}},\n",
    "        mappings={\n",
    "            \"properties\": {\n",
    "                \"text_embedding\": {\n",
    "                    \"type\": \"dense_vector\",\n",
    "                    \"dims\": 1024,\n",
    "                    \"element_type\": \"byte\",\n",
    "                },\n",
    "                \"text\": {\"type\": \"text\"},\n",
    "                \"id\": {\"type\": \"integer\"},\n",
    "                \"title\": {\"type\": \"text\"},\n",
    "            }\n",
    "        },\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acd3afe6",
   "metadata": {},
   "source": [
    "## Create the inference pipeline\n",
    "\n",
    "Now you have an inference endpoint and an index ready to store embeddings. The next\n",
    "step is to create an ingest pipeline that creates the embeddings using the\n",
    "inference endpoint and stores them in the index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "cec6f5c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ObjectApiResponse({'acknowledged': True})"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "client.ingest.put_pipeline(\n",
    "    id=\"cohere_embeddings\",\n",
    "    description=\"Ingest pipeline for Cohere inference.\",\n",
    "    processors=[\n",
    "        {\n",
    "            \"inference\": {\n",
    "                \"model_id\": \"cohere_embeddings\",\n",
    "                \"input_output\": {\n",
    "                    \"input_field\": \"text\",\n",
    "                    \"output_field\": \"text_embedding\",\n",
    "                },\n",
    "            }\n",
    "        }\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c5e1e7b",
   "metadata": {},
   "source": [
    "## Prepare data and insert documents\n",
    "\n",
    "This example uses the [SciFact](https://huggingface.co/datasets/mteb/scifact) data\n",
    "set that you can find on HuggingFace."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c71b8367",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Successfully loaded 20 documents\n",
      "Data ingestion completed, text embeddings generated!\n"
     ]
    }
   ],
   "source": [
    "#url = \"https://huggingface.co/datasets/mteb/scifact/raw/main/corpus.jsonl\"\n",
    "\n",
    "# Fetch the JSONL data from the URL\n",
    "#response = requests.get(url)\n",
    "#response.raise_for_status()  # Ensure we notice bad responses\n",
    "\n",
    "import json\n",
    "\n",
    "with open('./corpus.jsonl', 'r') as file:\n",
    "    content = file.read()\n",
    " \n",
    "# Split the content by new lines and parse each line as JSON\n",
    "data = [json.loads(line) for line in content.strip().split(\"\\n\") if line]\n",
    "\n",
    "data = data[:20]\n",
    "print(f\"Successfully loaded {len(data)} documents\")\n",
    "\n",
    "# Change `_id` key to `id` as `_id` is a reserved key in Elasticsearch.\n",
    "for item in data:\n",
    "    if \"_id\" in item:\n",
    "        item[\"id\"] = item.pop(\"_id\")\n",
    "\n",
    "# Prepare the documents to be indexed\n",
    "documents = []\n",
    "for line in data:\n",
    "    data_dict = line\n",
    "    documents.append(\n",
    "        {\n",
    "            \"_index\": \"cohere-embeddings\",\n",
    "            \"_source\": data_dict,\n",
    "        }\n",
    "    )\n",
    "\n",
    "# Use the bulk endpoint to index\n",
    "helpers.bulk(client, documents)\n",
    "\n",
    "print(\"Data ingestion completed, text embeddings generated!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9332078a",
   "metadata": {},
   "source": [
    "Your index is populated with the SciFact data and text embeddings for the text\n",
    "field."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610035af",
   "metadata": {},
   "source": [
    "## Hybrid search\n",
    "\n",
    "Let's start querying the index!\n",
    "\n",
    "The code below performs a hybrid search. The `kNN` query computes the relevance\n",
    "of search results based on vector similarity using the `text_embedding` field.\n",
    "The lexical search query uses BM25 retrieval to compute keyword similarity on\n",
    "the `title` and `text` fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ab735ef0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Title: BC1 RNA, the transcript from a master gene for ID element amplification, is able to prime its own reverse transcription.\n",
      "Text: ID elements are short interspersed elements (SINEs) found in high copy number in many rodent genomes. BC1 RNA, an ID-related transcript, is derived from the single copy BC1 RNA gene. The BC1 RNA gene has been shown to be a master gene for ID element amplification in rodent genomes. ID elements are dispersed through a process termed retroposition. The retroposition process involves a number of potential regulatory steps. These regulatory steps may include transcription in the appropriate tissue, transcript stability, priming of the RNA transcript for reverse transcription and integration. This study focuses on priming of the RNA transcript for reverse transcription. BC1 RNA gene transcripts are shown to be able to prime their own reverse transcription in an efficient intramolecular and site-specific fashion. This self-priming ability is a consequence of the secondary structure of the 3'-unique region. The observation that a gene actively amplified throughout rodent evolution makes a RNA capable of efficient self-primed reverse transcription strongly suggests that self-priming is at least one feature establishing the BC1 RNA gene as a master gene for amplification of ID elements.\n",
      "\n",
      "Title: The human myelin basic protein gene is included within a 179-kilobase transcription unit: expression in the immune and central nervous systems.\n",
      "Text: Two human Golli (for gene expressed in the oligodendrocyte lineage)-MBP (for myelin basic protein) cDNAs have been isolated from a human oligodendroglioma cell line. Analysis of these cDNAs has enabled us to determine the entire structure of the human Golli-MBP gene. The Golli-MBP gene, which encompasses the MBP transcription unit, is approximately 179 kb in length and consists of 10 exons, seven of which constitute the MBP gene. The human Golli-MBP gene contains two transcription start sites, each of which gives rise to a family of alternatively spliced transcripts. At least two Golli-MBP transcripts, containing the first three exons of the gene and one or more MBP exons, are produced from the first transcription start site. The second family of transcripts contains only MBP exons and produces the well-known MBPs. In humans, RNA blot analysis revealed that Golli-MBP transcripts were expressed in fetal thymus, spleen, and human B-cell and macrophage cell lines, as well as in fetal spinal cord. These findings clearly link the expression of exons encoding the autoimmunogen/encephalitogen MBP in the central nervous system to cells and tissues of the immune system through normal expression of the Golli-MBP gene. They also establish that this genetic locus, which includes the MBP gene, is conserved among species, providing further evidence that the MBP transcription unit is an integral part of the Golli transcription unit and suggest that this structural arrangement is important for the genetic function and/or regulation of these genes.\n",
      "\n",
      "Title: Arteriolar function in visceral adipose tissue is impaired in human obesity.\n",
      "Text: OBJECTIVE The purpose of this study was to characterize the relationship between adipose tissue phenotype and depot-specific microvascular function in fat. METHODS AND RESULTS In 30 obese subjects (age 42Â±11 years, body mass index 46Â±11 kg/m(2)) undergoing bariatric surgery, we intraoperatively collected visceral and subcutaneous adipose tissue and characterized depot-specific adipose phenotypes. We assessed vasomotor function of the adipose microvasculature using videomicroscopy of small arterioles (75-250 Î¼m) isolated from different fat compartments. Endothelium-dependent, acetylcholine-mediated vasodilation was severely impaired in visceral arterioles, compared to the subcutaneous depot (P<0.001 by ANOVA). Nonendothelium dependent responses to papaverine and nitroprusside were similar. Endothelial nitric oxide synthase inhibition with N(Ï‰)-nitro-l-arginine methyl ester reduced subcutaneous vasodilation but had no effect on severely blunted visceral arteriolar responses. Visceral fat exhibited greater expression of proinflammatory, oxidative stress-related, hypoxia-induced, and proangiogenic genes; increased activated macrophage populations; and had a higher capacity for cytokine production ex vivo. CONCLUSIONS Our findings provide clinical evidence that the visceral microenvironment may be intrinsically toxic to arterial health providing a potential mechanism by which visceral adiposity burden is linked to atherosclerotic vascular disease. Our findings also support the evolving concept that both adipose tissue quality and quantity may play significant roles in shaping cardiovascular phenotypes in human obesity.\n",
      "\n",
      "Title: The DNA Methylome of Human Peripheral Blood Mononuclear Cells\n",
      "Text: DNA methylation plays an important role in biological processes in human health and disease. Recent technological advances allow unbiased whole-genome DNA methylation (methylome) analysis to be carried out on human cells. Using whole-genome bisulfite sequencing at 24.7-fold coverage (12.3-fold per strand), we report a comprehensive (92.62%) methylome and analysis of the unique sequences in human peripheral blood mononuclear cells (PBMC) from the same Asian individual whose genome was deciphered in the YH project. PBMC constitute an important source for clinical blood tests world-wide. We found that 68.4% of CpG sites and <0.2% of non-CpG sites were methylated, demonstrating that non-CpG cytosine methylation is minor in human PBMC. Analysis of the PBMC methylome revealed a rich epigenomic landscape for 20 distinct genomic features, including regulatory, protein-coding, non-coding, RNA-coding, and repeat sequences. Integration of our methylome data with the YH genome sequence enabled a first comprehensive assessment of allele-specific methylation (ASM) between the two haploid methylomes of any individual and allowed the identification of 599 haploid differentially methylated regions (hDMRs) covering 287 genes. Of these, 76 genes had hDMRs within 2 kb of their transcriptional start sites of which >80% displayed allele-specific expression (ASE). These data demonstrate that ASM is a recurrent phenomenon and is highly correlated with ASE in human PBMCs. Together with recently reported similar studies, our study provides a comprehensive resource for future epigenomic research and confirms new sequencing technology as a paradigm for large-scale epigenomics studies.\n",
      "\n",
      "Title: Targeting A20 Decreases Glioma Stem Cell Survival and Tumor Growth\n",
      "Text: Glioblastomas are deadly cancers that display a functional cellular hierarchy maintained by self-renewing glioblastoma stem cells (GSCs). GSCs are regulated by molecular pathways distinct from the bulk tumor that may be useful therapeutic targets. We determined that A20 (TNFAIP3), a regulator of cell survival and the NF-kappaB pathway, is overexpressed in GSCs relative to non-stem glioblastoma cells at both the mRNA and protein levels. To determine the functional significance of A20 in GSCs, we targeted A20 expression with lentiviral-mediated delivery of short hairpin RNA (shRNA). Inhibiting A20 expression decreased GSC growth and survival through mechanisms associated with decreased cell-cycle progression and decreased phosphorylation of p65/RelA. Elevated levels of A20 in GSCs contributed to apoptotic resistance: GSCs were less susceptible to TNFalpha-induced cell death than matched non-stem glioma cells, but A20 knockdown sensitized GSCs to TNFalpha-mediated apoptosis. The decreased survival of GSCs upon A20 knockdown contributed to the reduced ability of these cells to self-renew in primary and secondary neurosphere formation assays. The tumorigenic potential of GSCs was decreased with A20 targeting, resulting in increased survival of mice bearing human glioma xenografts. In silico analysis of a glioma patient genomic database indicates that A20 overexpression and amplification is inversely correlated with survival. Together these data indicate that A20 contributes to glioma maintenance through effects on the glioma stem cell subpopulation. Although inactivating mutations in A20 in lymphoma suggest A20 can act as a tumor suppressor, similar point mutations have not been identified through glioma genomic sequencing: in fact, our data suggest A20 may function as a tumor enhancer in glioma through promotion of GSC survival. A20 anticancer therapies should therefore be viewed with caution as effects will likely differ depending on the tumor type.\n",
      "\n",
      "Title: Empirical Bayesian models for analysing molecular serotyping microarrays\n",
      "Text: BACKGROUND Microarrays offer great potential as a platform for molecular diagnostics, testing clinical samples for the presence of numerous biomarkers in highly multiplexed assays. In this study applied to infectious diseases, data from a microarray designed for molecular serotyping of Streptococcus pneumoniae was used, identifying the presence of any one of 91 known pneumococcal serotypes from DNA extracts. This microarray incorporated oligonucleotide probes for all known capsular polysaccharide synthesis genes and required a statistical analysis of the microarray intensity data to determine which serotype, or combination of serotypes, were present within a sample based on the combination of genes detected. RESULTS We propose an empirical Bayesian model for calculating the probabilities of combinations of serotypes from the microarray data. The model takes into consideration the dependencies between serotypes, induced by genes they have in common, and by homologous genes which, although not identical, are similar to each other in sequence. For serotypes which are very similar in capsular gene composition, extra probes are included on the microarray, providing additional information which is integrated into the Bayesian model. For each serotype combination with high probability, a second model, a Bayesian random effects model is applied to determine the relative abundance of each serotype. CONCLUSIONS To assess the accuracy of the proposed analysis we applied our methods to experimental data from samples containing individual serotypes and samples containing combinations of serotypes with known levels of abundance. All but two of the known serotypes of S. pneumoniae that were tested as individual samples could be uniquely determined by the Bayesian model. The model also enabled the presence of combinations of serotypes within samples to be determined. Serotypes with very low abundance within a combination of serotypes can be detected (down to 2% abundance in this study). As well as detecting the presence of serotype combinations, an approximate measure of the percentage abundance of the serotypes within the combination can be obtained.\n",
      "\n",
      "Title: Induction of myelodysplasia by myeloid-derived suppressor cells.\n",
      "Text: Myelodysplastic syndromes (MDS) are age-dependent stem cell malignancies that share biological features of activated adaptive immune response and ineffective hematopoiesis. Here we report that myeloid-derived suppressor cells (MDSC), which are classically linked to immunosuppression, inflammation, and cancer, were markedly expanded in the bone marrow of MDS patients and played a pathogenetic role in the development of ineffective hematopoiesis. These clonally distinct MDSC overproduce hematopoietic suppressive cytokines and function as potent apoptotic effectors targeting autologous hematopoietic progenitors. Using multiple transfected cell models, we found that MDSC expansion is driven by the interaction of the proinflammatory molecule S100A9 with CD33. These 2 proteins formed a functional ligand/receptor pair that recruited components to CD33â€™s immunoreceptor tyrosine-based inhibition motif (ITIM), inducing secretion of the suppressive cytokines IL-10 and TGF-Î² by immature myeloid cells. S100A9 transgenic mice displayed bone marrow accumulation of MDSC accompanied by development of progressive multilineage cytopenias and cytological dysplasia. Importantly, early forced maturation of MDSC by either all-trans-retinoic acid treatment or active immunoreceptor tyrosine-based activation motifâ€“bearing (ITAM-bearing) adapter protein (DAP12) interruption of CD33 signaling rescued the hematologic phenotype. These findings indicate that primary bone marrow expansion of MDSC driven by the S100A9/CD33 pathway perturbs hematopoiesis and contributes to the development of MDS.\n",
      "\n",
      "Title: The sacroiliac joint in the spondyloarthropathies.\n",
      "Text: The term spondyloarthropathy (SpA) describes and defines a group of related inflammatory joint disease that share characteristic clinical features and a unique association with the major histocompatibility complex class I molecule HLA-B27. Five subgroups can be differentiated: ankylosing spondylitis, reactive arthritis, psoriatic arthritis, arthritis associated with inflammatory bowel disease, and undifferentiated SpA. The sacroiliac joints are centrally involved in the SpA, most clearly and pathognomonic in ankylosing spondylitis, in which most patients are affected early in the disease. Overcoming some of the diagnostic difficulties of early sacroiliitis, dynamic magnetic resonance imaging was shown to visualize both acute and chronic changes in the sacroiliac joints. The inflammation in the sacroiliac joints in patients with SpA was recently examined in more detail; using immunohistology and in situ hybridrization, T cells, macrophages, and various cytokines were found in infiltrates. Biopsy specimens were obtained under guided computed tomography, and in the same study, intra-articular corticosteroid treatment was successfully undertaken. Further investigation of such biopsy specimens showed the absence of DNA of reactive arthritis-associated bacteria. The pathogenesis of the SpA and the reason for the tropism for the sacroiliac joints is still obscure. The nature of the relation of the genetic background of SpA to initially triggering bacterial infections remains to be established. In chronic disease, autoimmune mechanisms might be more important.\n",
      "\n",
      "Title: Bayesian measures of model complexity and fit\n",
      "Text: Summary. We consider the problem of comparing complex hierarchical models in which the number of parameters is not clearly defined. Using an information theoretic argument we derive a measure pD for the effective number of parameters in a model as the difference between the posterior mean of the deviance and the deviance at the posterior means of the parameters of interest. In general pD approximately corresponds to the trace of the product of Fisher's information and the posterior covariance, which in normal models is the trace of the â€˜hatâ€™ matrix projecting observations onto fitted values. Its properties in exponential families are explored. The posterior mean deviance is suggested as a Bayesian measure of fit or adequacy, and the contributions of individual observations to the fit and complexity can give rise to a diagnostic plot of deviance residuals against leverages. Adding pD to the posterior mean deviance gives a deviance information criterion for comparing models, which is related to other information criteria and has an approximate decision theoretic justification. The procedure is illustrated in some examples, and comparisons are drawn with alternative Bayesian and classical proposals. Throughout it is emphasized that the quantities required are trivial to compute in a Markov chain Monte Carlo analysis.\n",
      "\n",
      "Title: A rapid method for extraction of cotton (Gossypium spp.) genomic DNA suitable for RFLP or PCR analysis\n",
      "Text: Extraction of high-quality genomic DNA fromGossypium (cotton) species is difficult due to high levels of polysaccharide, oxidizable quinones, and other interfering substances. We describe a procedure that consistently permits isolation of cotton genomic DNA of satisfactory size and quality for RFLP and PCR analysis, as well as for most routine cloning applications. Several antioxidants, phenol-binding reagents, and phenol oxidase inhibitors are used throughout the procedure, and most polysaccharides are eliminated early in the procedure by isolation of nuclei.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "query = \"What is biosimilarity?\"\n",
    "\n",
    "response = client.search(\n",
    "    index=\"cohere-embeddings\",\n",
    "    size=100,\n",
    "    knn={\n",
    "        \"field\": \"text_embedding\",\n",
    "        \"query_vector_builder\": {\n",
    "            \"text_embedding\": {\n",
    "                \"model_id\": \"cohere_embeddings\",\n",
    "                \"model_text\": query,\n",
    "            }\n",
    "        },\n",
    "        \"k\": 10,\n",
    "        \"num_candidates\": 50,\n",
    "    },\n",
    "    query={\"multi_match\": {\"query\": query, \"fields\": [\"text\", \"title\"]}},\n",
    ")\n",
    "\n",
    "raw_documents = response[\"hits\"][\"hits\"]\n",
    "\n",
    "# Display the first 10 results\n",
    "for document in raw_documents[0:10]:\n",
    "    print(\n",
    "        f'Title: {document[\"_source\"][\"title\"]}\\nText: {document[\"_source\"][\"text\"]}\\n'\n",
    "    )\n",
    "\n",
    "# Format the documents for ranking\n",
    "documents = []\n",
    "for hit in response[\"hits\"][\"hits\"]:\n",
    "    documents.append(hit[\"_source\"][\"text\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153ce2d5",
   "metadata": {},
   "source": [
    "## Rerank search results\n",
    "\n",
    "To combine the results more effectively, use \n",
    "[Cohere's Rerank v3](https://docs.cohere.com/docs/rerank-2) model through the\n",
    "inference API to provide a more precise semantic reranking of the results.\n",
    "\n",
    "Create an inference endpoint with your Cohere API key and the used model name as\n",
    "the `model_id` (`rerank-english-v3.0` in this example)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "886c4d15",
   "metadata": {},
   "outputs": [],
   "source": [
    "client.inference.delete_model(inference_id=\"cohere_embeddings\")\n",
    "\n",
    "try:\n",
    "    client.inference.put_model(\n",
    "        task_type=\"rerank\",\n",
    "        inference_id=\"cohere_rerank\",\n",
    "        body={\n",
    "            \"service\": \"cohere\",\n",
    "            \"service_settings\": {\n",
    "                \"api_key\": COHERE_API_KEY,\n",
    "                \"model_id\": \"rerank-english-v3.0\",\n",
    "            },\n",
    "            \"task_settings\": {\n",
    "                \"top_n\": 10,\n",
    "            },\n",
    "        },\n",
    "    )\n",
    "except BadRequestError as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebf82d4a",
   "metadata": {},
   "source": [
    "Rerank the results using the new inference endpoint."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "296b6f2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Title: The sacroiliac joint in the spondyloarthropathies.\n",
      "Text: The term spondyloarthropathy (SpA) describes and defines a group of related inflammatory joint disease that share characteristic clinical features and a unique association with the major histocompatibility complex class I molecule HLA-B27. Five subgroups can be differentiated: ankylosing spondylitis, reactive arthritis, psoriatic arthritis, arthritis associated with inflammatory bowel disease, and undifferentiated SpA. The sacroiliac joints are centrally involved in the SpA, most clearly and pathognomonic in ankylosing spondylitis, in which most patients are affected early in the disease. Overcoming some of the diagnostic difficulties of early sacroiliitis, dynamic magnetic resonance imaging was shown to visualize both acute and chronic changes in the sacroiliac joints. The inflammation in the sacroiliac joints in patients with SpA was recently examined in more detail; using immunohistology and in situ hybridrization, T cells, macrophages, and various cytokines were found in infiltrates. Biopsy specimens were obtained under guided computed tomography, and in the same study, intra-articular corticosteroid treatment was successfully undertaken. Further investigation of such biopsy specimens showed the absence of DNA of reactive arthritis-associated bacteria. The pathogenesis of the SpA and the reason for the tropism for the sacroiliac joints is still obscure. The nature of the relation of the genetic background of SpA to initially triggering bacterial infections remains to be established. In chronic disease, autoimmune mechanisms might be more important.\n",
      "\n",
      "Title: Empirical Bayesian models for analysing molecular serotyping microarrays\n",
      "Text: BACKGROUND Microarrays offer great potential as a platform for molecular diagnostics, testing clinical samples for the presence of numerous biomarkers in highly multiplexed assays. In this study applied to infectious diseases, data from a microarray designed for molecular serotyping of Streptococcus pneumoniae was used, identifying the presence of any one of 91 known pneumococcal serotypes from DNA extracts. This microarray incorporated oligonucleotide probes for all known capsular polysaccharide synthesis genes and required a statistical analysis of the microarray intensity data to determine which serotype, or combination of serotypes, were present within a sample based on the combination of genes detected. RESULTS We propose an empirical Bayesian model for calculating the probabilities of combinations of serotypes from the microarray data. The model takes into consideration the dependencies between serotypes, induced by genes they have in common, and by homologous genes which, although not identical, are similar to each other in sequence. For serotypes which are very similar in capsular gene composition, extra probes are included on the microarray, providing additional information which is integrated into the Bayesian model. For each serotype combination with high probability, a second model, a Bayesian random effects model is applied to determine the relative abundance of each serotype. CONCLUSIONS To assess the accuracy of the proposed analysis we applied our methods to experimental data from samples containing individual serotypes and samples containing combinations of serotypes with known levels of abundance. All but two of the known serotypes of S. pneumoniae that were tested as individual samples could be uniquely determined by the Bayesian model. The model also enabled the presence of combinations of serotypes within samples to be determined. Serotypes with very low abundance within a combination of serotypes can be detected (down to 2% abundance in this study). As well as detecting the presence of serotype combinations, an approximate measure of the percentage abundance of the serotypes within the combination can be obtained.\n",
      "\n",
      "Title: Induction of myelodysplasia by myeloid-derived suppressor cells.\n",
      "Text: Myelodysplastic syndromes (MDS) are age-dependent stem cell malignancies that share biological features of activated adaptive immune response and ineffective hematopoiesis. Here we report that myeloid-derived suppressor cells (MDSC), which are classically linked to immunosuppression, inflammation, and cancer, were markedly expanded in the bone marrow of MDS patients and played a pathogenetic role in the development of ineffective hematopoiesis. These clonally distinct MDSC overproduce hematopoietic suppressive cytokines and function as potent apoptotic effectors targeting autologous hematopoietic progenitors. Using multiple transfected cell models, we found that MDSC expansion is driven by the interaction of the proinflammatory molecule S100A9 with CD33. These 2 proteins formed a functional ligand/receptor pair that recruited components to CD33â€™s immunoreceptor tyrosine-based inhibition motif (ITIM), inducing secretion of the suppressive cytokines IL-10 and TGF-Î² by immature myeloid cells. S100A9 transgenic mice displayed bone marrow accumulation of MDSC accompanied by development of progressive multilineage cytopenias and cytological dysplasia. Importantly, early forced maturation of MDSC by either all-trans-retinoic acid treatment or active immunoreceptor tyrosine-based activation motifâ€“bearing (ITAM-bearing) adapter protein (DAP12) interruption of CD33 signaling rescued the hematologic phenotype. These findings indicate that primary bone marrow expansion of MDSC driven by the S100A9/CD33 pathway perturbs hematopoiesis and contributes to the development of MDS.\n",
      "\n",
      "Title: A rapid method for extraction of cotton (Gossypium spp.) genomic DNA suitable for RFLP or PCR analysis\n",
      "Text: Extraction of high-quality genomic DNA fromGossypium (cotton) species is difficult due to high levels of polysaccharide, oxidizable quinones, and other interfering substances. We describe a procedure that consistently permits isolation of cotton genomic DNA of satisfactory size and quality for RFLP and PCR analysis, as well as for most routine cloning applications. Several antioxidants, phenol-binding reagents, and phenol oxidase inhibitors are used throughout the procedure, and most polysaccharides are eliminated early in the procedure by isolation of nuclei.\n",
      "\n",
      "Title: The DNA Methylome of Human Peripheral Blood Mononuclear Cells\n",
      "Text: DNA methylation plays an important role in biological processes in human health and disease. Recent technological advances allow unbiased whole-genome DNA methylation (methylome) analysis to be carried out on human cells. Using whole-genome bisulfite sequencing at 24.7-fold coverage (12.3-fold per strand), we report a comprehensive (92.62%) methylome and analysis of the unique sequences in human peripheral blood mononuclear cells (PBMC) from the same Asian individual whose genome was deciphered in the YH project. PBMC constitute an important source for clinical blood tests world-wide. We found that 68.4% of CpG sites and <0.2% of non-CpG sites were methylated, demonstrating that non-CpG cytosine methylation is minor in human PBMC. Analysis of the PBMC methylome revealed a rich epigenomic landscape for 20 distinct genomic features, including regulatory, protein-coding, non-coding, RNA-coding, and repeat sequences. Integration of our methylome data with the YH genome sequence enabled a first comprehensive assessment of allele-specific methylation (ASM) between the two haploid methylomes of any individual and allowed the identification of 599 haploid differentially methylated regions (hDMRs) covering 287 genes. Of these, 76 genes had hDMRs within 2 kb of their transcriptional start sites of which >80% displayed allele-specific expression (ASE). These data demonstrate that ASM is a recurrent phenomenon and is highly correlated with ASE in human PBMCs. Together with recently reported similar studies, our study provides a comprehensive resource for future epigenomic research and confirms new sequencing technology as a paradigm for large-scale epigenomics studies.\n",
      "\n",
      "Title: The suture provides a niche for mesenchymal stem cells of craniofacial bones\n",
      "Text: Bone tissue undergoes constant turnover supported by stem cells. Recent studies showed that perivascular mesenchymal stem cells (MSCs) contribute to the turnover of long bones. Craniofacial bones are flat bones derived from a different embryonic origin than the long bones. The identity and regulating niche for craniofacial-bone MSCs remain unknown. Here, we identify Gli1+ cells within the suture mesenchyme as the main MSC population for craniofacial bones. They are not associated with vasculature, give rise to all craniofacial bones in the adult and are activated during injury repair. Gli1+ cells are typical MSCs in vitro. Ablation of Gli1+ cells leads to craniosynostosis and arrest of skull growth, indicating that these cells are an indispensable stem cell population. Twist1(+/-) mice with craniosynostosis show reduced Gli1+ MSCs in sutures, suggesting that craniosynostosis may result from diminished suture stem cells. Our study indicates that craniofacial sutures provide a unique niche for MSCs for craniofacial bone homeostasis and repair.\n",
      "\n",
      "Title: BC1 RNA, the transcript from a master gene for ID element amplification, is able to prime its own reverse transcription.\n",
      "Text: ID elements are short interspersed elements (SINEs) found in high copy number in many rodent genomes. BC1 RNA, an ID-related transcript, is derived from the single copy BC1 RNA gene. The BC1 RNA gene has been shown to be a master gene for ID element amplification in rodent genomes. ID elements are dispersed through a process termed retroposition. The retroposition process involves a number of potential regulatory steps. These regulatory steps may include transcription in the appropriate tissue, transcript stability, priming of the RNA transcript for reverse transcription and integration. This study focuses on priming of the RNA transcript for reverse transcription. BC1 RNA gene transcripts are shown to be able to prime their own reverse transcription in an efficient intramolecular and site-specific fashion. This self-priming ability is a consequence of the secondary structure of the 3'-unique region. The observation that a gene actively amplified throughout rodent evolution makes a RNA capable of efficient self-primed reverse transcription strongly suggests that self-priming is at least one feature establishing the BC1 RNA gene as a master gene for amplification of ID elements.\n",
      "\n",
      "Title: Bayesian measures of model complexity and fit\n",
      "Text: Summary. We consider the problem of comparing complex hierarchical models in which the number of parameters is not clearly defined. Using an information theoretic argument we derive a measure pD for the effective number of parameters in a model as the difference between the posterior mean of the deviance and the deviance at the posterior means of the parameters of interest. In general pD approximately corresponds to the trace of the product of Fisher's information and the posterior covariance, which in normal models is the trace of the â€˜hatâ€™ matrix projecting observations onto fitted values. Its properties in exponential families are explored. The posterior mean deviance is suggested as a Bayesian measure of fit or adequacy, and the contributions of individual observations to the fit and complexity can give rise to a diagnostic plot of deviance residuals against leverages. Adding pD to the posterior mean deviance gives a deviance information criterion for comparing models, which is related to other information criteria and has an approximate decision theoretic justification. The procedure is illustrated in some examples, and comparisons are drawn with alternative Bayesian and classical proposals. Throughout it is emphasized that the quantities required are trivial to compute in a Markov chain Monte Carlo analysis.\n",
      "\n",
      "Title: induction of early IFN-inducible genes in the absence of type\n",
      "Text: On recognition of influenza virus (Flu) by TLR7, plasmacytoid dendritic cells (pDCs) produce type I IFN in significant amounts. Synthetic TLR7 ligands induce the maturation of pDCs, as evidenced by the expression of costimulatory molecules and the production of proinflammatory cytokines; however, they induce only low-level production of IFN-alpha. To dissect the TLR7 signaling in pDCs and how these different profiles are induced, we studied the effects of 2 TLR7 ligands (Flu and CL097) on the activation of blood-isolated pDCs and the human GEN2.2 pDC cell line. Type I IFN production by pDCs correlates with differential interferon regulatory factor 7 (IRF7) translocation into the nucleus induced by the 2 TLR7 ligands. Surprisingly, with both activators we nevertheless observed the rapid expression of the IFN-inducible genes mxa, cxcl10, and trail within 4 hours of stimulation. This expression, controlled by STAT1 phosphorylation, was independent of type I IFN. STAT1 activation was found to be strictly dependent on the PI3K-p38MAPK pathway, showing a new signaling pathway leading to rapid expression of IFN-inducible genes after TLR7 triggering. Thus, pDCs, through this unusual TLR7 signaling, have the capacity to promptly respond to viral infection during the early phases of the innate immune response.\n",
      "\n",
      "Title: Efficient targeting of expressed and silent genes in human ESCs and iPSCs using zinc-finger nucleases\n",
      "Text: Realizing the full potential of human embryonic stem cells (hESCs) and induced pluripotent stem cells (hiPSCs) requires efficient methods for genetic modification. However, techniques to generate cell typeâ€“specific lineage reporters, as well as reliable tools to disrupt, repair or overexpress genes by gene targeting, are inefficient at best and thus are not routinely used. Here we report the highly efficient targeting of three genes in human pluripotent cells using zinc-finger nuclease (ZFN)â€“mediated genome editing. First, using ZFNs specific for the OCT4 (POU5F1) locus, we generated OCT4-eGFP reporter cells to monitor the pluripotent state of hESCs. Second, we inserted a transgene into the AAVS1 locus to generate a robust drug-inducible overexpression system in hESCs. Finally, we targeted the PITX3 gene, demonstrating that ZFNs can be used to generate reporter cells by targeting non-expressed genes in hESCs and hiPSCs.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "response = client.inference.inference(\n",
    "    inference_id=\"cohere_rerank\",\n",
    "    body={\n",
    "        \"query\": query,\n",
    "        \"input\": documents,\n",
    "        \"task_settings\": {\"return_documents\": False},\n",
    "    },\n",
    ")\n",
    "\n",
    "# Reconstruct the input documents based on the index provided in the rereank response\n",
    "ranked_documents = []\n",
    "for document in response.body[\"rerank\"]:\n",
    "    ranked_documents.append(\n",
    "        {\n",
    "            \"title\": raw_documents[int(document[\"index\"])][\"_source\"][\"title\"],\n",
    "            \"text\": raw_documents[int(document[\"index\"])][\"_source\"][\"text\"],\n",
    "        }\n",
    "    )\n",
    "\n",
    "# Print the top 10 results\n",
    "for document in ranked_documents[0:10]:\n",
    "    print(f\"Title: {document['title']}\\nText: {document['text']}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9539ff47",
   "metadata": {},
   "source": [
    "## Retrieval Augmented Generation (RAG) with Cohere and Elasticsearch\n",
    "\n",
    "RAG is a method for generating text using additional information fetched from an\n",
    "external data source. With the ranked results, you can build a RAG system on\n",
    "top of what you created with \n",
    "[Cohere's Chat API](https://docs.cohere.com/docs/chat-api).\n",
    "\n",
    "Pass in the retrieved documents and the query to receive a grounded response\n",
    "using Cohere's newest generative model \n",
    "[Command R+](https://docs.cohere.com/docs/command-r-plus).\n",
    "\n",
    "Then pass in the query and the documents to the Chat API, and print out the\n",
    "response."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "39badebf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query: What is biosimilarity?\n",
      "Response: Sorry, I do not have any information about biosimilarity. Can I help you with something else?\n"
     ]
    }
   ],
   "source": [
    "response = co.chat(message=query, documents=ranked_documents, model=\"command-r-plus\")\n",
    "\n",
    "#source_documents = []\n",
    "#for citation in response.citations:\n",
    "#    for document_id in citation.document_ids:\n",
    "#        if document_id not in source_documents:\n",
    "#            source_documents.append(document_id)\n",
    "\n",
    "print(f\"Query: {query}\")\n",
    "print(f\"Response: {response.text}\")\n",
    "#print(\"Sources:\")\n",
    "#for document in response.documents:\n",
    "#    if document[\"id\"] in source_documents:\n",
    "#        print(f\"{document['title']}: {document['text']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ffc883e-3311-4ac1-969f-33054fc23764",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
